### DISCUSSION TITLE
Support for Content Security Policy

### DISCUSSION THREAD

**Discussion by Baroshem** [nuxt/nuxt#30404] [RFCs]
## Summary

The `contentSecurityPolicy` is a new top level configuration option in the `nuxt.config.ts` file that allows developers to easily configure [Content Security Policy](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) in their Nuxt apps.

## Implementation

Majority of the implementation will be taken from the Nuxt Security module like support for global plugins for handling nonce, and generating meta tag for SSG. To promote good security practices, the `contentSecurityPolicy` will come with a solid default option that should help majority of the users to have more secure applications. The default value will be taken from https://nuxt-security.vercel.app/headers/csp#strict-csp

Apart from just giving users ability to easily configure Content Security Policy header in SSR apps, this new addition will offer following features:
1. `nonce` generation -> https://nuxt-security.vercel.app/advanced/strict-csp#nonces
2. `subresource integrity` -> https://nuxt-security.vercel.app/utils/subresource-integrity
3. SSG `meta http-equiv` -> https://nuxt-security.vercel.app/advanced/strict-csp#csp-headers
4. Script and Style hashes -> https://nuxt-security.vercel.app/advanced/strict-csp#hashes
5. Export content security policy header to Nitro preset -> https://nuxt-security.vercel.app/advanced/strict-csp#csp-headers

## Inspirations

This RFC is inspired by the following sources:

1. https://nuxt-security.vercel.app/headers/csp
6. https://nextjs.org/docs/pages/building-your-application/configuring/content-security-policy

## API

In most cases, `contentSecurityPolicy` will not require any modification from the users as it should help make Nuxt apps more secure by default. But as all things in the Nuxt ecosystem, to deliver a great DX we have to support full customisability of the option to allow developers to modify it they way they need.

The example configuration could look like following:

```ts
export default defineNuxtConfig({
  contentSecurityPolicy: {
    value: {
    'script-src': [
        "'self'",  // Fallback value, will be ignored by most modern browsers (level 3)
        "https:", // Fallback value, will be ignored by most modern browsers (level 3)
        "'unsafe-inline'", // Fallback value, will be ignored by almost any browser (level 2)
        "'strict-dynamic'", // Strict CSP via 'strict-dynamic', supported by most modern browsers (level 3)
        "'nonce-{{nonce}}'" // Enables CSP nonce support for scripts in SSR mode, supported by almost any browser (level 2)
      ],
      'style-src': [
        "'self'", // Enables loading of stylesheets hosted on same origin
        "https:", // For increased security, replace by the specific hosting domain or file name of your external stylesheets
        "'unsafe-inline'" // Recommended default for most Nuxt apps
      ],
      'base-uri': ["'none'"],
      'img-src': ["'self'", "data:"], // Add relevant https://... sources if you load images from external sources 
      'font-src': ["'self'", "https:", "data:"], //  For increased security, replace by the specific sources for fonts
      'object-src': ["'none'"],
      'script-src-attr': ["'none'"],
      'upgrade-insecure-requests': true
    },
    nonce: true, // https://nuxt-security.vercel.app/advanced/strict-csp#nonces
    sri: true, // https://nuxt-security.vercel.app/utils/subresource-integrity
    ssg: {
      meta: true, // Enables CSP as a meta tag in SSG mode
      hashScripts: true, // Enables CSP hash support for scripts in SSG mode
      hashStyles: false, // Disables CSP hash support for styles in SSG mode (recommended)
      exportToPresets: true // Export security headers to Nitro presets
    },
  }
})
```

**Comment by TheAlexLichter** [comment]
Thanks for the RFC 🔥

It'd be also good to have an option to provide values for the nonce generation, which e.g. [Angular is allowing too](https://angular.dev/best-practices/security#content-security-policy).

Also, how can we handle SRI & Nonce for scripts added via Nuxt scripts / `useHead` 🤔

**Reply by vejja** [reply]
> Thanks for the RFC 🔥
> 
> It'd be also good to have an option to provide values for the nonce generation, which e.g. [Angular is allowing too](https://angular.dev/best-practices/security#content-security-policy).

I'd respectfully disagree with Angular's approach here. It opens the door for people to try to generate their nonces themselves in a way that is not RFC-compliant.

 
> Also, how can we handle SRI & Nonce for scripts added via Nuxt scripts / `useHead` 🤔

As soon as we have `strict-dynamic`, we are covered with everything under `useHead` for CSP. 
SRI is not automatically covered, but here is how we deal with it for Nuxt Security:
- If the script is loaded from an external source, Nuxt Scripts provides the `integrity` attribute
- If the script is bundled, we hash it ourselves
The only case that is not covered is when the script is loaded from an external source and the provider doesn't provide an integrity hash. In which case there is no way to have SRI for that script anyways...

**Comment by Baroshem** [comment]
@vejja 

Would love to get your input here :)

**Comment by vejja** [comment]
My 2 cents based on issues we get to address most often:
- don't make route-based CSP rules. It doesn't work well with Nuxt because when navigation happens on the client-side the CSP headers are not refreshed. I'd recommend to enforce one single set of app-wide CSP rules.
- Cut down the customization capabilities to the bare minimum, and enforce `strict-dynamic` with nonce or hashes. The CSP spec is truly awful, e.g. nonces don't apply to all elements, nonce and hashes cancel whitelists (but not the same ones, mind you...), strict-dynamic overrides other directives, etc. 

We improved the adoption of Nuxt Security when we provided reasonable defaults that enforce the above recommendations. Issues always arise when users try to depart from these defaults.

**Reply by Baroshem** [reply]
Very good comments, thanks Sebastien! :)

I have modified the code sample to not use the route-based CSP rules. Let me know if that works better now :)

For the second point, would you be able to showcase how you would modify the configuration so it would be smaller and easier to mainain? I am not sure if I will be able to cut everything properly ;)

**Comment by moshetanzer** [comment]
Hey. Really like this as a default. Could be talking nonsense or this might be obvious. But this should def be disabled in dev since it would slow down hmr by a ton if I understand correctly.

**Reply by Baroshem** [reply]
Hey, thanks for the comment!

I am not sure if it should be disabled in dev. I understand the concern about HMR performance but users are supposed to test whether their CSP was set correctly? If it will be disabled in dev, then users will now about it in production.

Unless you mean to add a config property `enabled` that will be set to false in dev, but you could enable it manually to test the configuration. But this could result in a situation where users will enable CSP, not test it properly in dev and release it to production and in production the app will not work due to the CSP rules that would block the application images/scripts/styles etc.

But maybe there is other way that we could make it work? :)

**Comment by martinszeltins** [comment]
One question that came to mind was - how would you dynamically add a whitelisted website? If backend sends back a list of websites and you wanted to add them to `contentSecurityPolicy`? I don't see how you would do that.

**Reply by Baroshem** [reply]
Hey, thanks for the comment. If I get this correctly, I don't think it is possible with this approach. But maybe you have a proposition on how it could be modified to address this use case? :)

**Reply by martinszeltins** [reply]
How about a macro or a composable similar to `usePageMeta`?

**Reply by Baroshem** [reply]
> How about a macro or a composable similar to `usePageMeta`?

Hmm, I dont think this could work. To make majority of the configuration to work, we need to execute it on the server (like with the middleware) or in build time (with nitro plugin).

But maybe I am missing some context here

**Reply by martinszeltins** [reply]
What if you wanted to have it dynamic? How would you produce a dynamic value?

**Comment by mikekidder** [comment]
Maybe reading this wrong, but not sure if best to make *contentSecurityPolicy* a top level configuration option.
This is especially true if I need to add the NuxtSecurity module to set other properties such as:
   Referral-Policy or Strict-Transport-Policy
   
So just provide the same default configuration that NuxtSecurity provided under security -> headers?

``` js
export default defineNuxtConfig({
  security: {
    headers: {
      contentSecurityPolicy: {
        'script-src': [
          "'self'", "https:", "'unsafe-inline'", "'strict-dynamic'", "'nonce-{{nonce}}'"
        ],
        'style-src': [
          "'self'", "https:", "'unsafe-inline'" 
        ],
        'base-uri': ["'none'"],
        'img-src': ["'self'", "data:"], 
        'font-src': ["'self'", "https:", "data:"], 
        'object-src': ["'none'"],
        'script-src-attr': ["'none'"],
        'upgrade-insecure-requests': true
      },
    },  
    nonce: true, 
    sri: true, 
    ssg: {
      meta: true, 
      hashScripts: true, 
      hashStyles: false, 
      exportToPresets: true 
    }    
  }
})
```

Also, would very much like to set contentSecurityPolicy : false in development. 
Might just want to enable/test CSP in my CI/CD pipeline.

``` js
export default defineNuxtConfig({
  security: {
    headers: {
      contentSecurityPolicy: false
    },  
  }
})
```

Perhaps add a property under *devServer* to override default (like CORS):

``` js
export default defineNuxtConfig({
  devServer: {
      security: {
        strict: false   // for Dev only; aligns with NuxtSecurity module
      }   
  }
})
```

**Reply by Baroshem** [reply]
Hey, thanks for the comments and additional context of your use case!

The idea of this RFC is to move some of the Nuxt Security functionality inside the core framework so that it can be used by all Nuxt users and allow them to have more secure apps by default without a need to install the Nuxt Security module. For those who need additional security improvements such as other headers, CSRF, Basic Auth or others they should be using Nuxt Security imho.

I think that if you already use Nuxt Security, you won't need this CSP configuration in Nuxt config. (We would probably also need to support the case where people might have Nuxt Security installed and then it should take priority instead of nuxt CSP configuration).

So in general, I think the idea is not to replace one with the other but more about taking the best practices about CSP specifically from Nuxt Security and make it a core framework feature. So that if you need additional security features, you would install Nuxt Security (as you probably do right now :)

Let me know if that makes sense and thanks once again for taking the time to share your thoughts! ❤️

**Reply by adamdehaven** [reply]
If an app already has a very customized Nuxt Security module configuration, this new built-in functionality would need to respect it, or treat Nuxt Securuty as an override I think?

Otherwise you'd have to explicitly disable this new default functionality somehow and ensure they don't conflict. 

Having a way built-in to view the "evaluated rules" would be key so that you can tell which settings are taking effect.

**Comment by harlan-zw** [comment]
IMO the benefit of Nuxt implementing this in the core is to try and encourage end-user security adoption for those who are either not aware of security best practices or are not inclined to set it up (lazy, too difficult, low priority).

To solve that, we'd need it prompt enabled with a fairly open policy that isn't going to get in the way but will show users when they are doing something potentially dangerous.

Having it enabled by default without immediate visibility would cause a lot of confusion and be a bit too much in the way. Introducing new config API surface just complicates an already complicated header.

I would propose that Nuxt CLI offers a "recommended" or "basic" preset (as opposed to a "minimal") that handles CSP. An interactive init command for modules is [already in the works](https://github.com/nuxt/cli/issues/754) so I think this is a good time to consider this approach.

Edit: bad idea for implementation

**Reply by Baroshem** [reply]
I am not sure about routeRules as pointed out by Sebastien here -> https://github.com/nuxt/nuxt/discussions/30404#discussioncomment-11719737. The support for defining rules for each route leads to confusion on client side navigation and maybe it would be better to handle it with one global config?

But I am up for providing this as a preset with two approaches; `basic` and `strict` as we already do it in Nuxt Security.

The idea for security guides is also in work which will be based on Nuxt Security documentation.

What would be your opinion @danielroe?

**Reply by harlan-zw** [reply]
It's not route based, it's just configured using route rules as this is the only way we have to configure all routes in one go without setting up a middleware

**Reply by atinux** [reply]
One note for /** is that it applies them on /api routes too

**Reply by vejja** [reply]
And one additional note is that CSP should not apply to /api routes 
It’s a HTML header, it does not make sense to serve it on routes that do not produce HTML content

**Reply by vejja** [reply]
> It's not route based, it's just configured using route rules as this is the only way we have to configure all routes in one go without setting up a middleware

@harlan-zw the issue we are facing with CSP is that CSP does not depend on the route,  it should be only one set of rules for the entire application

**Reply by harlan-zw** [reply]
Hmm we almost need a mime-type rules config

**Reply by vejja** [reply]
Yes. And also for the record ‘unsafe-inline’ on script-src is considered weak. 
The most versatile way to set it is ‘strict-dynamic nonce’ in SSR apps and ‘strict-dynamic sha’ on pre-rendered pages.

**Reply by harlan-zw** [reply]
I do agree. I would still like to see more of a minimal solution out of the box and point people toward using full-fledged solutions.

Something like...

```ts
import { WeakContentSecurityPolicy } from 'nuxt/config'

export default defineNuxtConfig({
   mimeHeaders: {
     /**
      * Restricts your app content weakly.
      * @see https://nuxt.com/docs/guides/security 
      */ 
     'text/html': WeakContentSecurityPolicy
   }
})
```

However, this may be wrong and we may need to just bite the bullet on a full CSP implementation.

Technically it could be handled in Unhead as a plugin so I'll have a think about that also.

**Comment by TheAlexLichter** [comment]
Also leaving https://vite.dev/guide/features.html#content-security-policy-csp here! 😋

**Reply by Baroshem** [reply]
And https://astro.build/blog/astro-590/#experimental-content-security-policy-support by @harlan-zw

**Comment by mrjackwills** [comment]
@TheAlexLichter I am correct in thinking that the vite option `html.cspNonce`, as it stands, has no effect on the Nuxt output? I'm trying to inject a nonce into all `<script>` and `<link as="script">` tags in my SSG pages. This is then substituted for a random value on each request via nginx.

So far I've had to create a custom, somewhat janky, nitro plugin in order to do edit the html files using the "close" hook.
